lib/Sema/CSApply.cpp

```diff
@@ -7005,12 +7005,30 @@ Expr *ExprRewriter::finishApply(ApplyExpr *apply, Type openedType,
     return apply;
   }
 
-  // If the metatype value isn't a type expression, the user should reference
-  // '.init' explicitly, for clarity.
   if (!cs.isTypeReference(fn)) {
-    cs.TC.diagnose(apply->getArg()->getStartLoc(),
-                   diag::missing_init_on_metatype_initialization)
-      .fixItInsert(apply->getArg()->getStartLoc(), ".init");
+    auto fnType = fn->getType();
+    auto &TC = cs.TC;
+
+    // If this is an attempt to initialize existential type.
+    if (auto metaType = fnType->getAs<MetatypeType>()) {
+      auto instanceType = metaTy->getInstanceType();
+      if (instanceType->isExistentialType()) {
+        auto diagnostic = fn->isStaticallyDerivedMetatype()
+                              ? diag::construct_protocol_by_name
+                              : diag::construct_protocol_value;
+
+        TC.diagnose(fn->getLoc(), diagnostic, metaTy)
+            .highlight(fn->getSourceRange());
+        return nullptr;
```

Comment:
Take a look at the 'depth' calculation in openExistentialReference() in CSApply.cpp. I think getArgCount() needs to account for this invalid case of directly calling a metatype.